Update Logging

E-B-Smith opened the pull request:
* Changed all SDK NSLog statements to `BNCLogDebug` statements.	
* Moved `BNCURLForBranchDirectory()` out of the PreferenceHelper class to modularize functionality.
* Added an explicit BNCLogCloseLogFile() function.
* Initialize the log when the Branch class is loaded.
* Turn on the debug messages when `[Branch setDebug]` is called.
* Updated the unit tests.
* Added a `LogLevelInfo` log level.

BranchMacMini commented:
Result of [Integration 1](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/d85abcdfd92da370bf9c50f78564f63b/d85abcdfd92da370bf9c50f785650119)
---
*Duration*: 8 minutes and 2 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

BranchMacMini commented:
Result of [Integration 2](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/d85abcdfd92da370bf9c50f78564f63b/d85abcdfd92da370bf9c50f78585572a)
---
*Duration*: 4 minutes and 4 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

ahmednawar commented on Branch-SDK/Branch-SDK/BNCDeviceInfo.m:
```diff
@@ -124,7 +125,7 @@ + (NSString*) userAgentString {
             preferences.browserUserAgentString = browserUserAgentString;
             preferences.lastSystemBuildVersion = self.systemBuildVersion;
-			//NSLog(@"[Branch] userAgentString: '%@'.", browserUserAgentString);
+			BNCLogDebug(@"userAgentString: '%@'.", browserUserAgentString);
```
@E-B-Smith  I don't fully understand the `BNCLog` code but my understanding is that methods like `BNCLogDebug` and `BNCLogWarning` don't check for flags like `isDebug` or `suppressLogWarnings`. Is this correct? If so, does this line mean that this log message will show up in all scenarios even in production apps that don't have `setDebug` enabled?

E-B-Smith replied:
The logging works like this:  The logger always writes the line to the Branch log file (along with data like the date and log level).  The `BNCLogSetLogDisplayLevel` function controls what shows in the console.  When `[Branch setDebug]` is called, it calls `BNCLogSetLogDisplayLevel()` so that all messages are shown in the console log.  Otherwise only warnings and errors are shown in the console.

ahmednawar replied:
Ok. Thanks for clarifying. What about the `suppressWarningLogs`? AFAICT, it is not checked when logging warnings. Am I missing something?

E-B-Smith replied:
Good catch.  I added support for the `suppressWarningLogs` flag.

aaustin commented:
Hang on for this one. Please don't merge until I can provide complete feedback.

aaustin commented on Branch-SDK/Branch-SDK/BNCDeviceInfo.m:
```diff
@@ -169,9 +170,9 @@ + (NSString*) userAgentString {
+                BNCLogDebug(@"Will userAgent.");
                 setBrowserUserAgent();
-                //NSLog(@"Did  userAgent.");
+                BNCLogDebug(@"Did  userAgent.");
```
Not sure these are useful logs. What's the purpose?

aaustin replied:
Also, if we're logging something for other people, we should definitely be way more descriptive in the message.

aaustin commented on Branch-SDK/Branch-SDK/BNCDeviceInfo.m:
```diff
@@ -124,7 +125,7 @@ + (NSString*) userAgentString {
             BNCPreferenceHelper *preferences = [BNCPreferenceHelper preferenceHelper];
             preferences.browserUserAgentString = browserUserAgentString;
             preferences.lastSystemBuildVersion = self.systemBuildVersion;
-			//NSLog(@"[Branch] userAgentString: '%@'.", browserUserAgentString);
```
I don't see why we'd need to log the user agent. This will be in our server logs

E-B-Smith replied:
It just shows up in the debug logs -- I find logging in real time makes much easier to debugging.  It's easier to pick up timing issues, for instance.

aaustin replied:
We recommend developers enable setDebug while integrating Branch. setDebug is not an internal debugging flag. We should only be logging messages that are useful to external developers and the user agent string is meaningless.

E-B-Smith replied:
Some of the debug message are useful for debugging the SDK and some are useful for external developers who are integrating Branch and debugging their own apps.

I can add another log level for internal use:  `BNCLogLevelDebugSDK`.

aaustin commented on Branch-SDK/Branch-SDK/BNCPreferenceHelper.m:
```diff
@@ -166,19 +167,15 @@ - (void) dealloc {
 #pragma mark - Debug methods
 
 - (void)log:(NSString *)filename line:(int)line message:(NSString *)format, ... {
```
can we get rid of this altogether if we're migrating logs?

E-B-Smith replied:
Yes, It's probably better to remove it completely and cut over to the new interface. 
Done.

BranchMacMini commented:
Result of [Integration 3](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/d85abcdfd92da370bf9c50f78564f63b/d85abcdfd92da370bf9c50f785ee041e)
---
*Duration*: 4 minutes and 9 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

BranchMacMini commented:
Result of [Integration 4](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/d85abcdfd92da370bf9c50f78564f63b/05401afc7daf59f84c1e3e6f6c0f6244)
---
*Duration*: 4 minutes and 8 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

aaustin commented on Branch-SDK/Branch-SDK/BNCServerInterface.m:
```diff
@@ -218,11 +216,9 @@ - (NSURLRequest *)preparePostRequest:(NSDictionary *)params url:(NSString *)url
-        [self.preferenceHelper log:FILE_NAME line:LINE_NUM message:@"body = %@", preparedParams];
-    }
+
+    BNCLogDebug(@" URL: %@.", url);
```
extra space here compared to line below

aaustin commented on Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m:
```diff
@@ -351,7 +352,7 @@ - (void) unloadViewController {
 - (void)safariViewController:(SFSafariViewController *)controller
       didCompleteInitialLoad:(BOOL)didLoadSuccessfully {
-    //NSLog(@"Safari Did load. Success: %d.", didLoadSuccessfully);   //  eDebug
+    BNCLogDebug(@"Safari Did load. Success: %d.", didLoadSuccessfully);
```
can we fix the capitalization here? Did -> did

aaustin commented on Branch-SDK/Branch-SDK/Branch.m:
```diff
@@ -1487,7 +1494,7 @@ - (void)processNextQueueItem {
             else if (![req isKindOfClass:[BranchOpenRequest class]] &&
                 (!self.preferenceHelper.deviceFingerprintID || !self.preferenceHelper.sessionID)) {
-                NSLog(@"[Branch Error] Missing session items!");
+                BNCLogError(@"Missing session items!");
```
We can probably remove this log. Impossible case and it's not the developer's fault.

aaustin commented on Branch-SDK/Branch-SDK/Branch.m:
```diff
@@ -1259,7 +1266,7 @@ - (NSString *)generateShortUrl:(NSArray *)tags
                 }
             }
-            NSLog(@"Branch SDK Error: making request before init succeeded!");
+            BNCLogError(@"Branch SDK Error: making request before init succeeded!");
```
Remove `Branch SDK Error` and capitalize M.

aaustin commented on Branch-SDK/Branch-SDK/Branch.m:
```diff
@@ -739,16 +750,12 @@ - (void)logoutWithCallback:(callbackWithStatus)callback {
-            if (self.preferenceHelper.isDebug) {
-                NSLog(@"Logout Success");
-            }
+            BNCLogDebug(@"Logout Success");
```
can you lower case the S in success and F in failure (right below)?

aaustin commented on Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m:
```diff
@@ -214,25 +215,25 @@ + (void) initialize {
     @synchronized (self) {
         if (!openRequestWaitQueueIsSuspended) {
-            //NSLog(@"Suspend openRequestWaitQueue.");
+            BNCLogDebug(@"Suspend openRequestWaitQueue.");
```
I'd argue that all of the logs in this class are meaningless to the developer and should not be here. They should have been deleted prior to feature release.

aaustin commented on Branch-SDK/Branch-SDK/BNCLog.m:
```diff
@@ -472,12 +473,19 @@ BNCLogOutputFunctionPtr _Nullable BNCLogOutputFunction() {
     }
 }
 
+void BNCLogCloseLogFile() {
```
You're not using this method anywhere. Why are you adding it?

E-B-Smith replied:
I use it in the unit tests and jdee uses it in his tests too.  It explicitly flushes closes log files.

BranchMacMini commented:
Result of [Integration 5](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/d85abcdfd92da370bf9c50f78564f63b/05401afc7daf59f84c1e3e6f6c805c9a)
---
*Duration*: 4 minutes and 9 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

E-B-Smith commented:
* Added the define `BNCLogDebugSDK` for displaying useful SDK debugging messages that aren't usually relevant for partners. These messages aren't shown when setDebug called, but we can turn it on when debugging the SDK.
* Fixed the punctuation and grammar for logging messages.
* Updated the unit tests.

aaustin commented:
Looks good to me 👍
